zephyrproject-rtos/zephyr
# tests: add CONFIG_NONDETERMINISTIC_TIMING

We have made great strides with the introduction of
CONFIG_QEMU_ICOUNT, but unfortunately this can't be applied to
every single emulator platform:

- icount isn't workable with SMP
- icount must be disabled for tests that interact with BT or
  networking
- some arches don't implement icount yet
- sometimes it is actually useful to test with icount disabled
  to introduce variance in test results, but these particular
  tests create noise to be sifted through

Introduce a configuration CONFIG_NONDETERMINISTIC_TIMING and
set it for QEMU targets that do not have icount enabled.
Conditionally compile test assertions, entire tests, or even
entire suites based on this, if they occasionally fail due to
host clock synchronization issues (and not some other reason).

Finding the tests that do this is a game of whack-a-mole but
this patch should get most of them. This will ultimately allow
us to stop auto-retrying tests in CI if they fail, which masks
other, real intermittent problems.

The scenarios that are excluded all either:
 - k_sleep() or k_busy_wait() while other events are expected
   to happen in between
 - call an API with a timeout that is not K_NO_WAIT or
   K_FOREVER and expect it to succeed within the provided
   timeout
 - use the system clock to measure a time interval between two
   events with some acceptable range defined

Filtering these makes tests that fail for some other reason
on these platforms much more obvious. There is still a great
deal of coverage and only a few platforms don't have icount
support.

This will not help with situations where the host is so
overloaded that tests can't complete within the timeout
defined by sanitycheck; sane values of -j should still be used.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>
